packages/next/src/lib/metadata/resolve-metadata.tsx

```diff
@@ -0,0 +1,215 @@
+        if (source.twitter && 'title' in source.twitter) {
+          updatedStashedTwitterTitle = resolveTitle(
+            titles.twitterTitle,
+            source.twitter.title
+          )
+          target.twitter.title = updatedStashedTwitterTitle
+        }
+      } else {
+        target.twitter = null
+      }
+    } else {
+      // TODO: Make sure the type is correct.
+      // @ts-ignore
+      target[key] = source[key]
+    }
+  }
+
+  return {
+    title: updatedStashedTitle,
+    openGraphTitle: updatedStashedOpenGraphTitle,
+    twitterTitle: updatedStashedTwitterTitle,
+  }
+}
+
+export async function resolveMetadata(metadataItems: Item[]) {
+  const resolvedMetadata = createDefaultMetadata()
+
+  const committedTitle: AbsoluteTemplateString = {
+    absolute: '',
+    template: '%s',
+  }
+  const committedOpenGraphTitle: AbsoluteTemplateString = {
+    absolute: '',
+    template: '%s',
+  }
+  const committedTwitterTitle: AbsoluteTemplateString = {
+    absolute: '',
+    template: '%s',
+  }
+  let stashedTitles
+
+  for (let i = 0; i < metadataItems.length; i++) {
+    const item = metadataItems[i]
+    if (item.type === 'layout' || item.type === 'page') {
+      let layerMod = await item.mod()
+
+      // Layer is a client component, we just skip it. It can't have metadata
+      // exported. Note that during our SWC transpilation, it should check if
+      // the exports are valid and give specific error messages.
+      if (
+        '$$typeof' in layerMod &&
+        (layerMod as any).$$typeof === Symbol.for('react.module.reference')
+      ) {
+        continue
+      }
+
+      if (layerMod.metadata && layerMod.generateMetadata) {
+        throw new Error(
+          'It is not allowed to export both `metadata` and `generateMetadata`. File: ' +
+            item.path
+        )
+      }
+
+      if (layerMod.metadata) {
+        stashedTitles = merge(layerMod.metadata, resolvedMetadata, {
+          title: committedTitle,
+          openGraphTitle: committedOpenGraphTitle,
+          twitterTitle: committedTwitterTitle,
+        })
+      } else if (layerMod.generateMetadata) {
+        // TODO: handle `generateMetadata`
+      }
+    }
+
+    // If we resolved all items in this layer, commit the stashed titles.
+    if (
+      stashedTitles &&
+      (i + 1 === metadataItems.length ||
+        metadataItems[i + 1].layer !== item.layer)
+    ) {
+      Object.assign(committedTitle, stashedTitles.title)
+      Object.assign(committedOpenGraphTitle, stashedTitles.openGraphTitle)
+      Object.assign(committedTwitterTitle, stashedTitles.twitterTitle)
+    }
+  }
+
+  return resolvedMetadata
+}
+
+// Generate the acutal React elements from the resolved metadata.
+export function elementsFromResolvedMetadata(metadata: ResolvedMetadata) {
+  return (
+    <>
+      {elementsFromResolvedBasic(metadata)}
+      {elementsFromResolvedAlternates(metadata)}
+      {metadata.openGraph !== null
```

Comment:
maybe put this `metadata.openGraph !== null` checking condition inside `elementsFromResolvedOpenGraph` like others?